File: Algorithms/src/main/java/com/threads/ThreadMultiJoin.java
package com.threads;

import java.util.Random;

/**
 * See threads/thread_multijoin.c from "The Linux Programming Interface" book
 */
public class ThreadMultiJoin {

	enum ThreadStatus { ALIVE, TERMINATED, JOINED }		
	
	
	//Structure to track thread status
	static class ThreadTracking {
		final Thread thread;
		ThreadStatus status;
		
		public ThreadTracking(Thread thread) {
			this.status = ThreadStatus.ALIVE;	
			this.thread = thread;
		}				
	}
	
	//Worker thread
	class Worker implements Runnable {
		final long sleep;
		final int pos;
		
		public Worker(long sleep, int pos) {
			this.sleep = sleep;
			this.pos = pos;
		}

		@Override
		public void run() {
			try {
				System.out.printf("Worker %2d sleeping millis %d%n", pos, sleep);
				Thread.sleep(sleep);										
				
				synchronized (lock) { //mark terminated and notify
					trackings[pos].status = ThreadStatus.TERMINATED;
					numUnjoined ++;
					lock.notify();	
				}				
			} catch (InterruptedException ie) {
				Thread.currentThread().interrupt();
			}			
		}
	}
	
	
	final int totalThreads;		
	
	final long[] sleeps;
	
	final ThreadTracking[] trackings; //access with lock
	int numLive; //alive or terminated
	int numUnjoined = 0; //terminated but not joined. Access with lock
	
	final Object lock = new Object();
	
	public ThreadMultiJoin(long[] sleeps) {
		this.sleeps = sleeps;
		totalThreads = sleeps.length;		
		trackings = new ThreadTracking[totalThreads];
		numLive = totalThreads;
	}
	
	public void run() throws InterruptedException {				
		//create threads
		System.out.printf("Creating threads: %d%n", totalThreads);
		for (int i = 0; i < totalThreads; i ++) {							
			Thread t = new Thread(new Worker(sleeps[i], i));
			trackings[i] = new ThreadTracking(t);						
			t.start();						
		}
				
		while (numLive > 0) { //iterate till we join all threads
			synchronized (lock) {
				while (numUnjoined == 0) { //wait till notification that some thread died
					lock.wait();
				}
				
				System.out.printf("Main notified on thread termination. NumUnjoined: %2d, numLive: %2d%n", numUnjoined, numLive);
				
				for (int i = 0; i < trackings.length; i ++) {
					ThreadTracking track = trackings[i];
					if (track.status == ThreadStatus.TERMINATED) {
						System.out.printf("Joining thread: %2d%n", i);
						track.thread.join(); //join thread
						
						track.status = ThreadStatus.JOINED;
						numLive --;
						numUnjoined --;															
					}
				}
			}
		}
		
		System.out.println("Done!");
	}

	public static void main(String[] args) throws Exception {
		System.out.printf("Available processors: %d%n", Runtime.getRuntime().availableProcessors());
		
		final int size = 1000;
		long[] sleeps = new long[size];
		Random r = new Random(System.currentTimeMillis());
		for (int i = 0; i < size; i ++) {
			sleeps[i] = (r.nextInt(5) + 1) * 1000;
		}				
				
		ThreadMultiJoin run = new ThreadMultiJoin(sleeps);
		run.run();
	}

}

File: c/Hello/src/thread_multijoin.c
#include "common.h"
#include <pthread.h>

enum tstate {
	ALIVE, TERMINATED, JOINED
};

static struct {
	pthread_t tid;
	enum tstate state;
	int slp;

} *threads;



pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t threadDiedCondition = PTHREAD_COND_INITIALIZER;


static int numLive; //live or terminated, but not joined. Access with lock
static int totalThreads; //constant
static int numUnjoined = 0; //terminated but not joined. Access with lock

static void* threadFunc(void* param) {
	int idx = (int) param;
	int s;

	printf("Worker %2d sleeping secs %d\n", idx, threads[idx].slp);

	s = sleep(threads[idx].slp); //sleep
	if (s != 0) {
		errExit("Worker: probably was interrupted");
	}

	s = pthread_mutex_lock(&lock); //lock
	if (s != 0) {
		errExit("Worker: pthread_mutex_lock");
	}

	numUnjoined ++;
	threads[idx].state = TERMINATED;

	s = pthread_mutex_unlock(&lock); //unlock
	if (s != 0) {
		errExit("Worker: pthread_mutex_unlock");
	}

	s = pthread_cond_signal(&threadDiedCondition); //notify
	if (s != 0) {
		errExit("Worker: pthread_cond_signal");
	}

	return NULL;
}

int main(int argc, char **argv) {
	int i;

	int params[100];
	int size = 100;
	for (i = 0; i < size; i ++) {
		params[i] = rand() % 5 + 1;
	}

	int s;
	numLive = totalThreads = size;

	//create threads
	printf("Creating threads: %d\n", totalThreads);
	threads = calloc(totalThreads, sizeof(*threads));
	if (threads == NULL) {
		errExit("Failed to allocate threads");
	}
	int slp;
	for (i = 0; i < size; i ++) {
		//slp = atoi(argv[i + 1]);
		slp = params[i];

		threads[i].slp = slp;
		threads[i].state = ALIVE;

		int idx = i;
		/*
		 * TODO: Important: see how parameter is given to func. Otherwise
		 * if I pass pointer, then variable in cycle will be overridden by further values in for-cycle.
		 */
		s = pthread_create(&threads[i].tid, NULL, threadFunc, (void*) idx);
		if (s != 0) {
			errExit("pthread_create");
		}
	}

	//join threads
	while (numLive > 0) {
		s = pthread_mutex_lock(&lock); //lock
		if (s != 0) {
			errExit("pthread_mutex_lock");
		}

		while (numUnjoined == 0) {
			s = pthread_cond_wait(&threadDiedCondition, &lock); //wait for notification
			if (s != 0) {
				errExit("pthread_cond_wait");
			}
		}

		printf("Main notified on thread termination. NumUnjoined: %2d, numLive: %2d\n", numUnjoined, numLive);

		for (i = 0; i < totalThreads; i ++) { //find terminated threads
			if (threads[i].state == TERMINATED) {
				printf("Joining thread: %2d\n", i);
				s = pthread_join(threads[i].tid, NULL);
				if (s != 0) {
					errExit("pthread_join");
				}

				threads[i].state = JOINED;
				numUnjoined --;
				numLive --;
			}
		}

		s = pthread_mutex_unlock(&lock); //lock
		if (s != 0) {
			errExit("pthread_mutex_unlock");
		}
	}

	printf("Done!\n");

	return 0;
}

